### ISSUE TITLE
Problem with variable length datatypes

### ISSUE THREAD

**Issue by classner** [h5py/h5py#862]
Dear h5py team,

I just ran into an issue concerning variable length datatypes. The following minimal example demonstrates the issue:

```
import numpy as np
import h5py
h5f = h5py.File('test.h5') 
varbindt = h5py.special_dtype(vlen=np.uint8)
dtype = np.dtype([('c1', varbindt), ('c2', varbindt), ('c3', varbindt)])
h5f.create_dataset('test3', dtype=dtype)
```
ValueError: Unable to insert member (Member overlaps with another member)

Curiously, it works with two columns, but not three or more. Am I missing out here something? For variable length strings, it works. I found #711 and #701, but didn't see the relation to this issue.

**Comment by aragilar** [comment]
Something weird is going on:
```python
>>> dtype = np.dtype([('c1', varbindt), ('c2', varbindt)])
>>> h5f.create_dataset('test3', dtype=dtype)
>>> h5f['test3']
<HDF5 dataset "test3": shape None, type "|V24">
>>> h5f['test3'].dtype
dtype({'names':['c1','c2'], 'formats':['O','O'], 'offsets':[0,16], 'itemsize':24})
>>> h5f['test3'].dtype['c1'].itemsize
8
```
and
```python
>>> dtype2 = np.dtype([('c1', varbindt), ('c2', varbindt)], align=True)
>>> dtype2
dtype([('c1', 'O'), ('c2', 'O')], align=True)
>>> t2 = h5t.py_create(dtype2)
>>> t2
<h5py.h5t.TypeCompoundID object at 0x7f6abf8c1780>
>>> t2.get_size()
16
>>> h5f.create_dataset('test2', dtype=dtype2)
Traceback (most recent call last):
  File "<input>", line 1, in <module>
    h5f.create_dataset('test2', dtype=dtype2)
  File "/home/aragilar/.virtualenvs/tmp-54046aa5da170e68/lib/python3.5/site-packages/h5py/_hl/group.py", line 108, in create_dataset
    dsid = dataset.make_new_dset(self, shape, dtype, data, **kwds)
  File "/home/aragilar/.virtualenvs/tmp-54046aa5da170e68/lib/python3.5/site-packages/h5py/_hl/dataset.py", line 97, in make_new_dset
    tid = h5t.py_create(dtype, logical=1)
  File "h5py/h5t.pyx", line 1543, in h5py.h5t.py_create (/tmp/pip-build-m39yqkt8/h5py/h5py/h5t.c:18118)
  File "h5py/h5t.pyx", line 1565, in h5py.h5t.py_create (/tmp/pip-build-m39yqkt8/h5py/h5py/h5t.c:17938)
  File "h5py/h5t.pyx", line 1587, in h5py.h5t.py_create (/tmp/pip-build-m39yqkt8/h5py/h5py/h5t.c:17416)
  File "h5py/h5t.pyx", line 1506, in h5py.h5t._c_compound (/tmp/pip-build-m39yqkt8/h5py/h5py/h5t.c:16484)
TypeError: Enforced alignment not compatible with HDF5 type
```
so I'm guessing something happens to variable types in compound types...

**Comment by aragilar** [comment]
Even more crazy:
```python
>>> varbindt2 = h5py.special_dtype(vlen=np.uint16)
>>> dtype3 = np.dtype([('c1', varbindt2), ('c2', varbindt)])
>>> h5f.create_dataset('test6', dtype=dtype3)
<HDF5 dataset "test6": shape None, type "|V24">
>>> h5f['test6'].dtype['c1'].itemsize
8
```

**Comment by linxd5** [comment]
> it works with two columns, but not three or more

I also meet this curious problem!

**Comment by linxd5** [comment]
It works for two columns:
```Python
import h5py
import numpy as np
dt = h5py.special_dtype(vlen=np.dtype('float32'))
data = np.array([(np.random.randn(3, 300), np.random.randn(30, 300)),
       (np.random.randn(32, 300), np.random.randn(3, 300))],
                dtype=np.dtype([("title", dt) ,("detail", dt)]))
f=h5py.File('foo.hdf5', 'w')
f.create_dataset('data', data=data)
f.close()
```
but not three:
```
import h5py
import numpy as np
dt = h5py.special_dtype(vlen=np.dtype('float32'))
data = np.array([(np.random.randn(3, 300), np.random.randn(30, 300), np.random.randn(30, 300)),
       (np.random.randn(4, 300), np.random.randn(31, 300), np.random.randn(29, 300))],
                dtype=np.dtype([("title", dt) ,("detail", dt), ("images", dt)]))
f=h5py.File('foo.hdf5', 'w')
f.create_dataset('data', data=data)
f.close()
```

**Comment by sethrj** [comment]
I'm seeing the same thing; it seems like a combination of a variable-length dtype (except for strings, strangely) and a logical. The first two datatypes are fine, and the third raises an exception:
```py
import numpy as np
import h5py

print(h5py.version.version)
print(h5py.version.hdf5_version)
print(np.version.version)

vidt = h5py.special_dtype(vlen=np.uint8)
f = h5py.File('temp.h5')

dt_good1 = np.dtype([
    ('foo', vidt),
    ('logical', np.bool)])
f.create_dataset('dt_good1',shape=(4,),dtype=dt_good1)

dt_good2 = np.dtype([
    ('foo', vidt),
    ('bar', vidt)])
f.create_dataset('dt_good2',shape=(4,),dtype=dt_good2)

dt_bad = np.dtype([
    ('foo', vidt),
    ('bar', vidt),
    ('logical', np.bool)])
f.create_dataset('dt_bad',shape=(4,),dtype=dt_bad)
```

Output:
```
2.7.0
1.10.1
1.13.1
Traceback (most recent call last):
  File "failure.py", line 25, in <module>
    f.create_dataset('dt_bad',shape=(4,),dtype=dt_bad)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/site-packages/h5py/_hl/group.py", line 108, in create_dataset
    dsid = dataset.make_new_dset(self, shape, dtype, data, **kwds)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/site-packages/h5py/_hl/dataset.py", line 97, in make_new_dset
    tid = h5t.py_create(dtype, logical=1)
  File "h5py/h5t.pyx", line 1543, in h5py.h5t.py_create (/opt/local/var/macports/build/_opt_bblocal_var_buildworker_ports_build_ports_python_py-h5py/py34-h5py/work/h5py-2.7.0/h5py/h5t.c:18112)
  File "h5py/h5t.pyx", line 1565, in h5py.h5t.py_create (/opt/local/var/macports/build/_opt_bblocal_var_buildworker_ports_build_ports_python_py-h5py/py34-h5py/work/h5py-2.7.0/h5py/h5t.c:17932)
  File "h5py/h5t.pyx", line 1587, in h5py.h5t.py_create (/opt/local/var/macports/build/_opt_bblocal_var_buildworker_ports_build_ports_python_py-h5py/py34-h5py/work/h5py-2.7.0/h5py/h5t.c:17410)
  File "h5py/h5t.pyx", line 1511, in h5py.h5t._c_compound (/opt/local/var/macports/build/_opt_bblocal_var_buildworker_ports_build_ports_python_py-h5py/py34-h5py/work/h5py-2.7.0/h5py/h5t.c:16594)
ValueError: Unable to insert member (Member overlaps with another member)
```

**Comment by sethrj** [comment]
Actually, is this fixed by 8647929ddfdf2e5af9b9621a1d2dd4064a477f16 ?

**Comment by sethrj** [comment]
OK, 
```py
h5py.h5t.py_create(dt_bad, logical=True)
```
fails but
```py
h5py.h5t.py_create(dt_bad, logical=False)
```
passes.

It looks like this is because `H5Tsize` of a vlen type is the size of two pointers, whereas the numpy size of an object is the size of a single pointer. The code in the current `_c_compound` function does not account for when unaligned compounds have members with different sizes from the numpy version. I'm submitting a pull request with a fix.
